### DISCUSSION TITLE
Iterate over PrismaClient keys to call PrismaAction

### DISCUSSION THREAD

**Discussion by chrishoermann** [prisma/prisma#5107] [Q&A]
### What i try to do: 

I want to create a function that iterates over an array of `DeleteElement` (code below) and executes a`PrismaAction` on the model defined in `DeleteElement` e.g delete all `userAddress`for the a `user` with a certain `id`;  

```typescript
import { PrismaClient } from ".prisma/client";
import { prisma } from "../../context";

export interface DeleteElement {
  dbKey: Exclude<
    keyof PrismaClient,
    | "$transaction"
    | "$on"
    | "$connect"
    | "$disconnect"
    | "$use"
    | "$executeRaw"
    | "$queryRaw"
  >;
  idKey: string;
  id: string;
}

export const iterateDelete = (elements: DeleteElement[]) => {
  try {
    elements.forEach(async (elem) => {
      await prisma[elem.dbKey].deleteMany({
        where: {
          [elem.idKey]: elem.id,
        },
      });
    });
  } catch (error) {
    console.log("Error: ", error);
    throw new Error("Error when deleting Elements from Database");
  }
};

// usage of function

iterateDelete([
  { dbKey: "userAddress", idKey: "userId", id: "INSERT ID HERE" },
  { dbKey: "userConfig", idKey: "userId", id: "INSERT ID HERE" },
]);
```

### Problem:

The `PrismaActions` are defined in the `Delegate` interfaces like for `UserAddress` in `UserAddressDelegate` 

```typescript
    deleteMany<T extends UserAddressDeleteManyArgs>(
      args?: Subset<T, UserAddressDeleteManyArgs>
    ): Promise<BatchPayload>
```

now typescript complains that 
`This expression is not callable.
  Each member of the union type '(<T extends CompanyDeleteManyArgs>(args?: Subset<T, CompanyDeleteManyArgs> | undefined) => Promise<BatchPayload>) | ... 7 more ... | (<T extends RegionDeleteManyArgs>(args?: Subset<...> | undefined) => Promise<...>)' has signatures, but none of those signatures are compatible with each other.`

This is kind of logical because each `deleteMany` has a unique `args` signature.

### Question

is it possible to iterate over models and call `PrismaActions` like this or am I headed in the wrong direction with this approach?
Any help is appreciated!
Thanks!

**Comment by timsuchanek** [comment]
Hi @chrishoermann, this looks quite advanced what you're doing here 😄 
I can already tell you, that without Prisma Client exposing additional types, this will get very hard.
A few questions:

Is delete just an example and you want other actions as well?
Is it a hard requirement for you to pass in `idKey` and `id` or could it also have same shape as the `deleteMany` input's `where` argument?

**Reply by chrishoermann** [reply]
Hi @timsuchanek!
thanks for the thorough reply. 

It's not a hard requirement to add in `idKey` and `id`. This was just the first thought that came to my mind when I thought about a solution. The time I read your idea of passing in the whole `where` argument I was like: "why didn't I think about this myself". 

For now I just want to implement `deleteMany` but maybe in the future other `Actions` would come in handy - don't know yet :smirk:

**Comment by timsuchanek** [comment]
I think to have typesafety within your `iterateDelete` function, you need to "discriminate" the types like so:
```ts
import { PrismaClient } from './@prisma/client'

const prisma = new PrismaClient()

type Model = 'user' | 'post'
type ModelAction =
  | 'findUnique'
  | 'findMany'
  | 'findFirst'
  | 'create'
  | 'update'
  | 'updateMany'
  | 'upsert'
  | 'delete'
  | 'deleteMany'
  | 'aggregate'
  | 'count'

type GetArgumentType<M extends Model, Action extends ModelAction> = Parameters<
  PrismaClient[M][Action]
>[0]

type DeleteOperation = {
  model: Model
  where: GetArgumentType<Model, 'delete'>['where']
}

async function runDeleteOperations<T extends DeleteOperation[]>(operations: T) {
  for (const operation of operations) {
    if (operation.model === 'user') {
      await prisma[operation.model].deleteMany({
        where: operation.where,
      })
    }
    if (operation.model === 'post') {
      await prisma[operation.model].deleteMany({
        where: operation.where,
      })
    }
  }
}

async function main() {
  await runDeleteOperations([
    {
      model: 'user',
      where: {
        id: 'asd',
      },
    },
    {
      model: 'post',
      where: {
        id: 'asd',
      },
    },
  ])
  prisma.$disconnect()
}

main().catch((e) => {
  console.error(e)
})
```
If you don't need type-safety in your delete function and just in its interface, you can use an `as any` in there.

The `Model` and `ModelAction` types don't exist right now, but we can add them if you need them. Just let me know.

**Reply by timsuchanek** [reply]
The reason you're seeing that error, is that you pass in a union, which can't clearly differentiate to which function it should go then. That's why the `if` statements are needed in my snippet. This is code you could generate.

**Reply by chrishoermann** [reply]
Aren't the `ModelAction` types already exposed in `.prisma\client` as `Prisma.PrismaAction`?

Having a pure `Model` type would come in handy although extracting the `Model` types from `PrismaClient` by excluding `$transaction, $on, ...` would also work (at least I think so :smirk:) - don't know if there are any downfalls with the latter approach or if it's to "hacky"? :smiley:

**Reply by chrishoermann** [reply]
But your solution looks great. :+1: I'll definitely play around with it and see where it takes me 😄

**Comment by Download** [comment]
```ts
import type { PrismaClient } from '@prisma/client';

type Models = {
  [K in keyof PrismaClient as K extends `$${string}` ? never : K]:  PrismaClient[K];
}

// given 3 models, user, password and note,
// `Models` looks like this:
// type Models = {
//     readonly user: Prisma.UserDelegate<Prisma.RejectOnNotFound | Prisma.RejectPerOperation | undefined>;
//     readonly password: Prisma.PasswordDelegate<...>;
//     readonly note: Prisma.NoteDelegate<...>;
// } 
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
